### ISSUE TITLE
Add attempt to log file naming

### ISSUE THREAD

**Issue by nick-youngblut** [snakemake/snakemake#122]
When using `--restart-times`, the log file is instantly deleted after an attempt fails, so the user cannot look at the log file to see what went wrong. The user must instead wait until the last attempt has finished. For long running jobs or a long cluster queue, this can be a slow process. I've previously waited for hours/days just to find out that I had a typo in my `shell` commands, which caused all attempts to fail. It would be great if 'attempt' could be added to the log file names in order to see the logs from each attempt.

**Comment by SichongP** [comment]
Are you referring to `log` directive? If that's the case, snakemake doesn't actually handle logging. You write to log files by specifically providing `{log}` to the programs you're using, or by redirecting output to log file. In either case the file will be overwritten by most programs when you run it again with same file path provided.

You can get around this by:
1. Add an `onerror` key word to move all your log files to a different location (say, `attempts/1/`) if workflow fails.
2. Alternatively, you can use an input function to name log files based on the time workflow is evaluated. This should give you different log files at each attempt.

**Comment by nick-youngblut** [comment]
Yes, I'm referring to `log:`. I would prefer to have log work like `input:` in which a function can be provided with `attempt` as one of the potential arguments. Thanks for the suggests, but they are both very hack-y

**Comment by SichongP** [comment]
Yeah I don't think `log` supports input function right now. It can already access wildcard names the same way `input` does and I don't think it needs that as of now. You can always use `params` which does support input function in place of `log`. 

I personally think `onerror` is a more organized way for your problem tho.

**Comment by nick-youngblut** [comment]
I guess that I could just hybridize `resources:` and `log:` to get what I want. Example:

```
rule Example:
    input:
        /path/to/input
    output:
       /path/to/output
    resources:
       attempt = lambda wildcards, attempt: '_{}.txt'.format(attempt)
    log:
        /path/to/log
    shell:
        """
        ./some_command {input} {output} 2> {log}{resources.attempt}
        """
```

I haven't tried it, but I'm wondering if that would work or whether snakemake will complain that the  `{log}` file cannot be found (instead the `{log}{resources.attempt}` is written).

**Comment by SultanOrazbayev** [comment]
I came across a similar problem on a cluster, so in these cases I looked at the cluster scheduler logs (they are saved with the unique job id, which is another potential solution to your problem, though in this case the number of logs can grow very large very quickly).

With your attempt, it might be better to put it in `params` directive (as suggested above), but looks like it could work (also didn't test it).

**Comment by SultanOrazbayev** [comment]
https://bitbucket.org/snakemake/snakemake/issues/420/onerror-onsuccess-for-rules

**Comment by SultanOrazbayev** [comment]
It turns out that attempt must be passed via `resources:`. I tried experimenting with shell variable expansions, but in the end that was useless because on every attempt, the original `log` file (without a suffix) will be erased.

In the end your solution works, I ran it `snakemake test --restart-times 3`, where the `Snakefile` is:
```
rule test:
    resources: attempt = lambda wildcards, attempt: attempt,
    output: 'test.txt'
    log: 'test.log'
    shell:
        '''
        echo {resources.attempt} > {log}{resources.attempt}
        '''
```
Note that `test.log` will always be empty.

**Comment by lczech** [comment]
:+1: from my side. Would be nice if

```
    ...
    log: "run-{attempt}.log"
    ...
```

worked.

**Comment by fgvieira** [comment]
But then we'd be treating `attempt` as an extra `wildcard` and, if so, we'd also have to support in `output`, `log` and `benchmark`, to maintain `snakemake` wildcard constraint, no?

**Comment by lczech** [comment]
Hm, as a wildcard for output file name, it is for sure useful for the log files, because otherwise it is impossible to see the error messages that caused earlier attempts to fail, if the log file gets overwritten each time.

As for the others (output, benchmark), it _might_ be useful to have as an available wildcard, for some edge use cases. But I don't know if it can then be left optional easily? If I understand snakemake file naming rules, the same wildcards have to be used across all those files then, right? That might not be the right call here. Usually, output should be overwritten with a new attempt, and also benchmarks of a failed attempt are usually not that meaningful. 

In other words, if adding `attempt` as a wildcard so that it _can_ be used for the log would then also _require_ to use it for the out and benchmark file names, then that seems not like the best solution. Maybe there is a way to treat it as a special variable somehow? Ugly of course, but in some sense, it is special, as it is not a regular wildcard that is actually determining the flow and connection of rules, and rather a repeat of the same rule.

**Comment by fgvieira** [comment]
Would it make more sense for the log to just be appended (since it is never deleted)?

**Comment by lczech** [comment]
Haha well that for sure solves the general issue, seems rather simple to implement, and does not have any added complexity for the rest. Yeah. Very good solution, I think! Simplicity beats everything!

**Comment by fgvieira** [comment]
But that would be up to the user, since the `log` behavior is defined in `shell`:
```
shell:
    "cmd > {log}"
```
or:
```
shell:
    "cmd >> {log}"
```

**Comment by lczech** [comment]
Would that work though? I vaguely remember that Snakemake internally deletes the log file for new attempts, but haven't tested this recently.
